StandaloneMmPkg/Library/PeiStandaloneMmHobProductionLib/PeiStandaloneMmHobProductionLib.c

```diff
@@ -0,0 +1,128 @@
+/** @file
+  This  library instance installs HOBs whose data can be acquired using Specification
+  defined interfaces in PEI. This data is consumed during the initialization of the
+  Standalone MM environment.
+
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Ppi/MpServices.h>
+#include <Guid/MpInformation.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/PeiServicesLib.h>
+
+/**
+  Notification function called when EFI_PEI_MP_SERVICES_PPI becomes available. This function produces
+  the MP Information HOB needed to initialize Standalone MM.
+
+  @param[in] PeiServices          Indirect reference to the PEI Services Table.
+  @param[in] NotifyDescriptor     Address of the notification descriptor data structure.
+  @param[in] Ppi                  Address of the PPI that was installed.
+
+  @retval   EFI_SUCCESS           The MP Info HOB was produced successfully.
+  @retval   EFI_OUT_OF_RESOURCES  Insufficient memory resources are available to allocate the HOB buffer.
+
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+CreateMpInfoHob (
+  IN EFI_PEI_SERVICES           **PeiServices,
+  IN EFI_PEI_NOTIFY_DESCRIPTOR  *NotifyDescriptor,
+  IN VOID                       *Ppi
+  )
+{
+  EFI_STATUS               Status;
+  UINTN                    NumOfCpus;
+  UINTN                    NumOfEnabledCpus;
+  UINTN                    MpInfoSize;
+  UINTN                    Index;
+  MP_INFORMATION_HOB_DATA  *HobData;
+  EFI_PEI_MP_SERVICES_PPI  *MpPpi;
+
+  NumOfCpus        = 0;
+  NumOfEnabledCpus = 0;
+
+  MpPpi = (EFI_PEI_MP_SERVICES_PPI *)Ppi;
+  ASSERT (MpPpi != NULL);
+
+  if (MpPpi == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = MpPpi->GetNumberOfProcessors ((CONST EFI_PEI_SERVICES **)PeiServices, MpPpi, &NumOfCpus, &NumOfEnabledCpus);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "[%a] Get number of processors %d, and number of enabled processors %d - %r.\n",
+      __FUNCTION__,

```

Comment:
If you make any changes, can you switch these to `__func__`?

Reply:
I'm doing that now.

Reply:
Done. Rebased and had to resolve conflicts due to https://github.com/microsoft/mu_basecore/commit/ece8b6cdf63218917f6db1b0d669c2079a631286.